File: Strings/GroupAnagrams.java
// LeetCode 49 - Group Anagrams
// Time Complexity: O(n * k log k) | Space Complexity: O(n * k)
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class GroupAnagrams {
    public List<List<String>> groupAnagrams(String[] strs) {
        Map<String, List<String>> groups = new HashMap<>();

        for (String s : strs) {
            char[] chars = s.toCharArray();
            Arrays.sort(chars);
            String key = String.valueOf(chars);
            groups.computeIfAbsent(key, value -> new ArrayList<>()).add(s);
        }

        return new ArrayList<>(groups.values());
    }
}

File: Strings/GroupAnagrams.go
// LeetCode 49 - Group Anagrams
// Time Complexity: O(n * k log k) | Space Complexity: O(n * k)
package main

import (
	"sort"
)

func groupAnagrams(strs []string) [][]string {
	groups := make(map[string][]string)

	for _, s := range strs {
		chars := []byte(s)
		sort.Slice(chars, func(i, j int) bool {
			return chars[i] < chars[j]
		})
		key := string(chars)
		groups[key] = append(groups[key], s)
	}

	answer := make([][]string, 0, len(groups))
	for _, group := range groups {
		answer = append(answer, group)
	}

	return answer
}
